mito-ai/mito_ai/handlers.py

```diff
@@ -26,12 +28,77 @@
     SmartDebugMessageMetadata,
     CodeExplainMessageMetadata,
     InlineCompletionMessageMetadata,
+    HistoryReply
 )
+from .prompt_builders import remove_inner_thoughts_from_message
 from .providers import OpenAIProvider
 from .utils.create import initialize_user
+from .utils.schema import MITO_FOLDER
 
 __all__ = ["CompletionHandler"]
 
+# Global message history with thread-safe access
+class GlobalMessageHistory:
+    def __init__(self, save_file: str = os.path.join(MITO_FOLDER, "message_history.json")):
+        self._lock = Lock()
+        self._llm_history: List[Dict[str, str]] = []
+        self._display_history: List[Dict[str, str]] = []
+        self._save_file = save_file
+
+        # Load from disk on startup
+        self._load_from_disk()
+    
+    def _load_from_disk(self):
+        """Load existing history from disk, if it exists."""
+        if os.path.exists(self._save_file):
+            try:
+                with open(self._save_file, "r", encoding="utf-8") as f:
+                    data = json.load(f)
+                    self._llm_history = data.get("llm_history", [])
+                    self._display_history = data.get("display_history", [])
+            except Exception as e:
+                print(f"Error loading history file: {e}")
+    
+    def _save_to_disk(self):
+        """Save current history to disk."""
+        data = {
+            "llm_history": self._llm_history,
+            "display_history": self._display_history,
+        }
+        # Using a temporary file and rename for safer "atomic" writes
+        tmp_file = f"{self._save_file}.tmp"
+        try:
+            with open(tmp_file, "w", encoding="utf-8") as f:
+                json.dump(data, f, indent=2)
+            os.replace(tmp_file, self._save_file)
+        except Exception as e:
+            # log or handle error
+            print(f"Error saving history file: {e}")
+
+    def get_histories(self) -> tuple[List[Dict[str, str]], List[Dict[str, str]]]:
+        with self._lock:
+            return self._llm_history[:], self._display_history[:]
+
+    def clear_histories(self) -> None:
+        with self._lock:
+            self._llm_history = []
+            self._display_history = []
+            self._save_to_disk()
+
+    def append_message(self, llm_message: Dict[str, str], display_message: Dict[str, str]) -> None:
+        with self._lock:
+            self._llm_history.append(llm_message)
+            self._display_history.append(display_message)
+            self._save_to_disk()
+
+    def truncate_histories(self, index: int) -> None:
```

Comment:
Two quick things here: 

1. It looks like we are not removing the original message that the user edits. Maybe just an off by one bug here? 
2. In the UI, when the user edits the message and presses submit, we are not clearing all future messages until the AI responds. That looks a bit whacky. Can we clear the messages right away. If this isn't a regression introduced by this PR, then feel free to just open an issue and we can fix independently. 


https://github.com/user-attachments/assets/e43c492b-33b3-4021-aedb-969f2e585b1e

Reply:
1. Yes, off by one bug. I re-added system message in the frontend, so the indexing starts from 1 on the frontend.
2. Looks like a regression introduced by the previous PR